dbms/src/Common/CachingAllocator.h

```diff
@@ -0,0 +1,875 @@
+#pragma once
+
+#include <atomic>
+#include <mutex>
+#include <list>
+#include <memory>
+#include <random>
+#include <pcg_random.hpp>
+#include <unordered_map>
+#include <sys/mman.h>
+#include <boost/intrusive/list.hpp>
+#include <boost/intrusive/set.hpp>
+#include <boost/noncopyable.hpp>
+#include <ext/scope_guard.h>
+
+#include <Common/Exception.h>
+#include <Common/randomSeed.h>
+#include <Common/formatReadable.h>
+
+/// Old Darwin builds lack definition of MAP_ANONYMOUS
+#ifndef MAP_ANONYMOUS
+#define MAP_ANONYMOUS MAP_ANON
+#endif
+
+namespace DB::ErrorCodes
+{
+extern const int CANNOT_ALLOCATE_MEMORY;
+extern const int CANNOT_MUNMAP;
+}
+
+struct CachingAllocatorStats
+{
+    size_t total_chunks_size {0};
+    size_t total_allocated_size {0};
+    size_t total_size_currently_initialized {0};
+    size_t total_size_in_use {0};
+
+    size_t num_chunks {0};
+    size_t num_regions {0};
+    size_t num_free_regions {0};
+    size_t num_regions_in_use {0};
+    size_t num_keyed_regions {0};
+
+    size_t hits {0};
+    size_t concurrent_hits {0};
+    size_t misses {0};
+
+    size_t allocations {0};
+    size_t allocated_bytes {0};
+    size_t evictions {0};
+    size_t evicted_bytes {0};
+    size_t secondary_evictions {0};
+};
+
+/**
+ * Thread-safe caching allocator interface.
```

Comment:
This needs more description. Looks like it is some kind of memory allocator, backed by mmap, that also uses the unused memory for some kind of cache. Probably it caches and allocates the same blocks of fixed size. The cache is supposed to be used (and evicted) from multiple threads, so the cache blocks probably use some internal reference counting? For allocated blocks, does it use the same internal reference counting, or does it require the caller to track that, like normal allocators do? What's the cache eviction algorithm? The description mentions memory fragmentation, does the evicition algorithm does something specific to address it? And so on.